lib/Parse/SyntaxParseActions.cpp

```diff
@@ -0,0 +1,170 @@
+        LeadingTrivia(LeadingTrivia), TrailingTrivia(TrailingTrivia),
+        Range(Range) {}
+};
+
+struct DeferredLayoutNode {
+  syntax::SyntaxKind Kind;
+  ArrayRef<RecordedOrDeferredNode> Children;
+  unsigned Length;
+
+  DeferredLayoutNode(syntax::SyntaxKind Kind,
+                     ArrayRef<RecordedOrDeferredNode> Children, unsigned Length)
+      : Kind(Kind), Children(Children), Length(Length) {}
+};
+
+OpaqueSyntaxNode SyntaxParseActions::makeDeferredToken(tok tokenKind,
+                                                       StringRef leadingTrivia,
+                                                       StringRef trailingTrivia,
+                                                       CharSourceRange range,
+                                                       bool isMissing) {
+  return new (DeferredNodeAllocator) DeferredTokenNode(
+      isMissing, tokenKind, leadingTrivia, trailingTrivia, range);
+}
+
+OpaqueSyntaxNode SyntaxParseActions::makeDeferredLayout(
+    syntax::SyntaxKind k, bool isMissing,
+    const ArrayRef<RecordedOrDeferredNode> &children) {
+  assert(!isMissing && "Missing layout nodes not implemented yet");
+
+  // Compute the length of this node.
+  unsigned length = 0;
+  for (auto &child : children) {
+    switch (child.getKind()) {
+    case RecordedOrDeferredNode::Kind::Null:
+      break;
+    case RecordedOrDeferredNode::Kind::Recorded:
+      llvm_unreachable("Children of deferred nodes must also be deferred");
+      break;
+    case RecordedOrDeferredNode::Kind::DeferredLayout:
+      length +=
+          static_cast<const DeferredLayoutNode *>(child.getOpaque())->Length;
+      break;
+    case RecordedOrDeferredNode::Kind::DeferredToken:
+      length += static_cast<const DeferredTokenNode *>(child.getOpaque())
+                    ->Range.getByteLength();
+      break;
+    }
+  }
+
+  return new (DeferredNodeAllocator) DeferredLayoutNode(k, children, length);
+}
+
+OpaqueSyntaxNode
+SyntaxParseActions::recordDeferredToken(OpaqueSyntaxNode deferred) {
+  auto Data = static_cast<const DeferredTokenNode *>(deferred);
+  if (Data->IsMissing) {
+    return recordMissingToken(Data->TokenKind, Data->Range.getStart());
+  } else {
+    return recordToken(Data->TokenKind, Data->LeadingTrivia,
+                       Data->TrailingTrivia, Data->Range);
+  }
+}
+
+OpaqueSyntaxNode
+SyntaxParseActions::recordDeferredLayout(OpaqueSyntaxNode deferred) {
+  auto Data = static_cast<const DeferredLayoutNode *>(deferred);
+
+  auto childrenStore =
+      DeferredNodeAllocator.Allocate<OpaqueSyntaxNode>(Data->Children.size());
+  MutableArrayRef<OpaqueSyntaxNode> children =
+      llvm::makeMutableArrayRef(childrenStore, Data->Children.size());
+
+  for (size_t i = 0; i < Data->Children.size(); ++i) {
+    auto Child = Data->Children[i];
+    switch (Child.getKind()) {
+    case RecordedOrDeferredNode::Kind::Null:
+    case RecordedOrDeferredNode::Kind::Recorded:
```

Comment:
Other methods states "Children of deferred nodes must also be deferred". So this should be unreachable, right?

Reply:
Yes, you’re correct.